beacon-chain/sync/pending_blocks_queue.go

```diff
@@ -49,81 +49,67 @@ func (s *Service) processPendingBlocksQueue() {
-		trace.Int64Attribute("numSlots", int64(len(ss))),
-		trace.Int64Attribute("numPeers", int64(len(pids))),
-	)
+	// Sort slots for ordered processing.
+	sortedSlots := s.sortedPendingSlots()
+
+	span.AddAttributes(trace.Int64Attribute("numSlots", int64(len(sortedSlots))), trace.Int64Attribute("numPeers", int64(len(s.cfg.p2p.Peers().Connected()))))
 
 	randGen := rand.NewGenerator()
-	for _, slot := range ss {
-		// process the blocks during their respective slot.
-		// otherwise wait for the right slot to process the block.
+	var parentRoots [][32]byte
+
+	// Iterate through sorted slots.
+	for _, slot := range sortedSlots {
+		// Skip processing if slot is in the future.
 		if slot > s.cfg.clock.CurrentSlot() {
 			continue
 		}
 
-		ctx, span := trace.StartSpan(ctx, "processPendingBlocks.InnerLoop")
-		span.AddAttributes(trace.Int64Attribute("slot", int64(slot))) // lint:ignore uintcast -- This conversion is OK for tracing.
+		ctx, span := startInnerSpan(ctx, slot)
 
-		s.pendingQueueLock.RLock()
-		bs := s.pendingBlocksInCache(slot)
-		// Skip if there's no block in the queue.
-		if len(bs) == 0 {
-			s.pendingQueueLock.RUnlock()
+		// Get blocks in the pending queue for the current slot.
+		blocksInCache := s.getBlocksInQueue(slot)
+		if len(blocksInCache) == 0 {
 			span.End()
 			continue
 		}
-		s.pendingQueueLock.RUnlock()
 
-		// Loop through the pending queue and mark the potential parent blocks as seen.
-		for _, b := range bs {
-			if b == nil || b.IsNil() || b.Block().IsNil() {
-				span.End()
+		// Process each block in the queue.
+		for _, b := range blocksInCache {
+			if err := blocks.BeaconBlockIsNil(b); err != nil {
 				continue
 			}
-
 			blkRoot, err := b.Block().HashTreeRoot()
 			if err != nil {
-				tracing.AnnotateError(span, err)
-				span.End()
 				return err
 			}
-			// No need to process the same block if we are already processing it
+
+			// Skip blocks that are already being processed.
 			if s.cfg.chain.BlockBeingSynced(blkRoot) {
-				rootString := fmt.Sprintf("%#x", blkRoot)
-				log.WithField("BlockRoot", rootString).Info("Skipping pending block already being processed")
+				log.WithField("BlockRoot", fmt.Sprintf("%#x", blkRoot)).Info("Skipping pending block already being processed")
 				continue
 			}
 
-			inDB := s.cfg.beaconDB.HasBlock(ctx, blkRoot)
-			// No need to process the same block twice.
-			if inDB {
-				s.pendingQueueLock.Lock()
-				if err = s.deleteBlockFromPendingQueue(slot, b, blkRoot); err != nil {
-					s.pendingQueueLock.Unlock()
+			// Remove and skip blocks already in the database.
+			if s.cfg.beaconDB.HasBlock(ctx, blkRoot) {
+				if err := s.removeBlockFromQueue(b, blkRoot); err != nil {
 					return err
 				}
-				s.pendingQueueLock.Unlock()
-				span.End()
 				continue
 			}
 
-			s.pendingQueueLock.RLock()
-			inPendingQueue := s.seenPendingBlocks[b.Block().ParentRoot()]
-			s.pendingQueueLock.RUnlock()
+			inPendingQueue := s.isBlockInQueue(b.Block().ParentRoot())
```

Comment:
parent root is used below on 122
`parentRoot := b.Block().ParentRoot()`
another small thing but maybe it can be used here before passing in maybe doesn't matter much